packages/@aws-cdk/cdk/lib/program.ts

```diff
@@ -0,0 +1,176 @@
+import cxapi = require('@aws-cdk/cx-api');
+import fs = require('fs');
+import path = require('path');
+import { IConstruct, Root } from './core/construct';
+import { merge } from './util/json';
+import { validateAndThrow } from './util/validation';
+
+/**
+ * Represents a CDK program.
+ */
+export class Program extends Root {
+  /**
+   * Return the default singleton Program instance
+   */
+  public static defaultInstance(): Program {
+    if (Program.DefaultInstance === undefined) {
+      Program.DefaultInstance = new Program();
+      Program.DefaultInstance.initializeDefaultProgram();
+    }
+    return Program.DefaultInstance;
+  }
+
+  private static DefaultInstance?: Program;
+
+  constructor() {
+    super();
+
+    this.loadContext();
+  }
+
+  /**
+   * Runs the program. Output is written to output directory as specified in the request.
+   */
+  public run(): void {
+    const outdir = process.env[cxapi.OUTDIR_ENV];
+    if (!outdir) {
+      process.stderr.write(`ERROR: The environment variable "${cxapi.OUTDIR_ENV}" is not defined\n`);
+      process.stderr.write('AWS CDK Toolkit (>= 0.11.0) is required in order to interact with this program.\n');
+      process.exit(1);
+      return;
+    }
+
+    const result = this.synthesizeAll();
+
+    const outfile = path.join(outdir, cxapi.OUTFILE_NAME);
+    fs.writeFileSync(outfile, JSON.stringify(result, undefined, 2));
+  }
+
+  /**
+   * Calculate the CDK app response
+   */
+  public synthesizeAll(): cxapi.SynthesizeResponse {
+    this.node.prepareTree();
+
+    // first, validate this stack and stop if there are errors.
+    validateAndThrow(this);
+
+    const result: any = {
+      version: cxapi.PROTO_RESPONSE_VERSION,
+      runtime: this.collectRuntimeInformation()
+    };
+
+    for (const synthesizable of this.node.findAll().filter(isSynthesizable)) {
+      merge(result, synthesizable.synthesize(), true);
+    }
+
+    return result;
+  }
+
+  /**
+   * Initialize the default Program instance
+   */
+  private initializeDefaultProgram() {
```

Comment:
I’d give this method a more specific name: “autoRunOnExit” or something

Reply:
I'd suggest simply `runBeforeExit`.